### DISCUSSION TITLE
Discussing the implementation of in-memory image-transfer.

### DISCUSSION THREAD

**Discussion by LI7XI** [LI7XI/rainmeter#4] [Ideas]

**Comment by LI7XI** [comment]
hey @d-uzlov, I was able to compile rainmeter locally (using visual studio, it didn't eat all ram this time), and somehow i succeeded in compiling a test plugin with clang on wsl.

But i still had no idea where to start in rainmeter src code, if you have time (only if you have time), i hope you help me in figuring this out.

**Comment by LI7XI** [comment]
I was staring at this method:

<details>

```cpp

bool MeterImage::Update()
{
	if (Meter::Update())
	{
		if (!m_Measures.empty() || m_DynamicVariables)
		{
			// Store the current values so we know if the image needs to be updated
			std::wstring oldResult = m_ImageNameResult;

			if (!m_Measures.empty())  // read from the measures
			{
				if (m_ImageName.empty())
				{
					m_ImageNameResult = m_Measures[0]->GetStringOrFormattedValue(AUTOSCALE_OFF, 1.0, 0, false);
				}
				else
				{
					m_ImageNameResult = m_ImageName;
					if (!ReplaceMeasures(m_ImageNameResult, AUTOSCALE_OFF))
					{
						// ImageName doesn't contain any measures, so use the result of MeasureName.
						m_ImageNameResult = m_Measures[0]->GetStringOrFormattedValue(AUTOSCALE_OFF, 1.0, 0, false);
					}
				}
			}
			else  // read from the skin
			{
				m_ImageNameResult = m_ImageName;
			}

			LoadImage(m_ImageNameResult, (wcscmp(oldResult.c_str(), m_ImageNameResult.c_str()) != 0));

			return true;
		}
		else if (m_NeedsRedraw)
		{
			m_NeedsRedraw = false;
			return true;
		}
	}
	return false;
}

```
</details>

If i got it correctly, it checks: if measure name || dynamic variables options are used, then read image from "measure name" string location, otherwise read from image name option.
I don't get what happened here: `if (m_ImageName.empty())`.

What i thought about is, in the `else` of the above if statement, we check if the measure can provide image data, if so, load the data from there. Otherwise use a string value.

For me it looks simple but how to check if a measure can provide a specific thing?
And after we get the image data, how to use it? i thought `LoadImage` is a function but it turned out to be a macro, so i really don't get how this entire `update` method works.

**Comment by d-uzlov** [comment]
Rainmeter code is so hard to read :(

> I don't get what happened here: `if (m_ImageName.empty())`.

Apparently, Rainmeter allows you to write something like this
```ini
ImageName=path/to/the/image/folder/[&MyMeasure]
MeasureName=AnotherMeasure
```
And the actual path to image will be `path/to/the/image/folder/[MyMeasure]`, and the `MeasureName` option will be ignored.
Well, I can't say I'm too surprised.

If I were allowed to, I would refactor it completely, but the most minimal changes I would make to support in-memory image transfer would be something like following:
<details>

```cpp
// Store the current values so we know if the image needs to be updated
std::wstring oldResult = m_ImageNameResult;

m_ImageNameResult = m_ImageName;

if (!m_Measures.empty())  // read from the measures
{
	if (m_ImageName.empty())
	{
		auto imageLoaded = LoadImageFromMeasure(m_Measures[0]);
		if (imageLoaded) {
			return true;
		}
		m_ImageNameResult = m_Measures[0]->GetStringOrFormattedValue(AUTOSCALE_OFF, 1.0, 0, false);
	}
	else
	{
		m_ImageNameResult = m_ImageName;
		if (!ReplaceMeasures(m_ImageNameResult, AUTOSCALE_OFF))
		{
			// ImageName doesn't contain any measures, so use the result of MeasureName.
			m_ImageNameResult = m_Measures[0]->GetStringOrFormattedValue(AUTOSCALE_OFF, 1.0, 0, false);
		}
	}
}

LoadImage(m_ImageNameResult, (wcscmp(oldResult.c_str(), m_ImageNameResult.c_str()) != 0));

return true;
```
</details>

Kinda ugly, but arguably you can't make it more ugly than it already was.

**Comment by d-uzlov** [comment]
> i thought `LoadImage` is a function but it turned out to be a macro, so i really don't get how this entire `update` method works.

I think I remember this shit. If I'm not mistaken, `LoadImage` is a method of image meter class, which _just happens to have name collision_ with Windows macro. Fortunately (or unfortunately, depending on how you look at it), the macro replaces method name both in definition and in usage places, so it compiles.

> For me it looks simple but how to check if a measure can provide a specific thing?

You will have to modify the plugin measure class.
Plugin measure class will probe the plugin dll to determine whether it supports images, and translate requests to load the image into the dll function, just like it does with `Update`, `GetString`, etc.

**Comment by LI7XI** [comment]
>Rainmeter code is so hard to read :(

yep :(

>If I were allowed to, I would refactor it completely, but the most miniature changes I would make to support in-memory image transfer would be something like following:

What is this function `LoadImageFromMeasure()`?

**Comment by d-uzlov** [comment]
> What is this function `LoadImageFromMeasure()`?

If should be a new method, similar to `LoadImage`, but querying the plugin for image data, instead of reading it from file.
Exact implementation would depend on plugin image transfer implementation.

**Comment by LI7XI** [comment]
>If should be a new method, similar to LoadImage, but querying the plugin for image data, instead of reading it from file.
Exact implementation would depend on plugin image transfer implementation.

ah.... i've been tracing all `LoadImage` related functions, duplicating and renaming to `LoadImageFromMeasure` `GeneralImage::LoadFromMeasure` etc.., i was planning to do necessary changes in order to make it read image data from memory.

Until i came across `D2DBitmapLoader`, my brain batteries ran out of charge because i need to at least rewrite half of it, most of it using handlers and stuff that involve writing to disk (if i got it correctly), and there is other functions like `GetFileInfo` which i really don't get the need of it in meters.

Now, unread all what i said and tell me how you were planning to implement this?😹

**Comment by LI7XI** [comment]
>ah.... i've been tracing all LoadImage related functions, duplicating and renaming to LoadImageFromMeasure GeneralImage::LoadFromMeasure etc.., i was planning to do necessary changes in order to make it read image data from memory.

I know i know this is terrible, but i just want to get it working, after that i will be more organized.

**Comment by d-uzlov** [comment]
> Until i came across `D2DBitmapLoader`, my brain batteries ran out of charge because i need to at least rewrite half of it, most of it using handlers and stuff that involve writing to disk (if i got it correctly), and there is other functions like `GetFileInfo` which i really don't get the need of it in meters.
> 
> Now, unread all what i said and tell me how you were planning to implement this?😹

Last time I looked at the rainmeter code was something like 3 years ago, before their big redesign. It was not that messy back then. Looks like they managed to fuck the code up much more than it was before. I'm shocked how spagetti it is.

Later I will post my version of how it can be relatively cheaply changed to accommodate for plugin images.

**Comment by d-uzlov** [comment]
> Until i came across `D2DBitmapLoader`, my brain batteries ran out of charge

I know that feel. Kinda frustrating.
I made few hints on how to handle this d2d mess: https://github.com/LI7XI/rainmeter/pull/2
They are just hints, I didn't test anything, but it seems like you don't have to rewrite anything, it would be enough to just add few functions here and there.

**Comment by d-uzlov** [comment]
> how you were planning

I was _planning_ that it wouldn't be so coupled with d2d stuff.

**Comment by LI7XI** [comment]
Gemme a sec, i'm learning about how to review pull requests, idk how to make inline comments lol

**Comment by d-uzlov** [comment]
To add an inline comment you open the "Files Changed" tab and click on the `+` symbol that appears on the left of the line when you hover your mouse cursor over a line.

**Comment by d-uzlov** [comment]
I don't have more time right now, I'll look at your comments tomorrow, if you add any.

**Comment by LI7XI** [comment]
>I don't have more time right now, I'll look at your comments tomorrow, if you add any.

Sure, thanks soooo much for your help ^^

**Comment by LI7XI** [comment]
@d-uzlov hehehehehehe, i'm so happy!!


https://user-images.githubusercontent.com/55055075/120841926-62396300-c56c-11eb-972d-f1a83b629be4.mp4

First of all, thanks sooo very much a lot, i wouldn't have done it without your help (≧∇≦)ﾉ

Now, what do you (as a plugin dev) think of the following?

<details>

```ini
[Rainmeter]
Update=1000

[ImageName]
Measure=Plugin
Plugin=RmPlugin

[Image]
Meter=Image
MeasureName=ImageName
W=300
PreserveAspectRatio=2
```

```cpp
#include <Windows.h>
#include "../lib/RmAPI/RainmeterAPI.h"

// add this struct to your rainmeterapi.h, you can use it here if you want

struct PluginImageData
{
	UINT8* pixels;
	UINT width;
	UINT height;
	UINT64 timestamp;
};

#include <cmath>
#include <chrono>

struct Measure
{
  Measure() {}

  PluginImageData imageData;
};

UINT8 random(UINT8 min, UINT8 max) //range : [min, max]
{
  static bool first = true;
  if (first)
  {
    srand(time(NULL)); //seeding for the first time only!
    first = false;
  }
  return min + rand() % ((max + 1) - min);
}

void UpdatePixels(PluginImageData &imageData)
{
  UINT8 rand_r = random(0, 255);
  UINT8 rand_g = random(0, 255);
  UINT8 rand_b = random(0, 255);
  for (UINT y = 0; y < imageData.height; ++y)
    for (UINT x = 0; x < imageData.width; ++x)
    {
      UINT8 r = rand_r;
      UINT8 g = (y < 150) ? 0 : rand_g;
      UINT8 b = (x < 200) ? 0 : rand_b;
      imageData.pixels[y * imageData.width * 4 + x * 4 + 0] = r;
      imageData.pixels[y * imageData.width * 4 + x * 4 + 1] = g;
      imageData.pixels[y * imageData.width * 4 + x * 4 + 2] = b;
      imageData.pixels[y * imageData.width * 4 + x * 4 + 3] = 255;
    }
}

PLUGIN_EXPORT void Initialize(void **data, void *rm)
{
  Measure *measure = new Measure;
  UINT width = 500;
  UINT height = 300;

  measure->imageData.pixels = new UINT8[width * height * 4]();
  measure->imageData.width = width;
  measure->imageData.height = height;
  measure->imageData.timestamp = 0;

  *data = measure;
}

PLUGIN_EXPORT void Reload(void *data, void *rm, double *maxValue)
{
  Measure *measure = (Measure *)data;
}

PLUGIN_EXPORT double Update(void *data)
{
  Measure *measure = (Measure *)data;
  measure->imageData.timestamp++;
  UpdatePixels(measure->imageData);
  return 0.0;
}

PLUGIN_EXPORT LPCWSTR GetImage(void *data, PluginImageData &imageData)
{
  Measure *measure = (Measure *)data;
  // auto imgData = measure->imageData;
  imageData.pixels = measure->imageData.pixels;
  imageData.width = measure->imageData.width;
  imageData.height = measure->imageData.height;
  imageData.timestamp = measure->imageData.timestamp;
  return L"";
}

PLUGIN_EXPORT void Finalize(void *data)
{
  Measure *measure = (Measure *)data;
  delete[] measure->imageData.pixels;
  delete measure;
}
```

</details>


I have few notes, i'm thinking about removing `timestamp` and in `LoadBitmapFromPluginMeasure` function (i changed some names) we set timestamp to `chrono clock.now()` or something, technically it should work.

Also idk somewhere in a function that loads the bitmap, i saw a format like `...r8g8b8a8`, so i think rainmeter uses `uint8` type for pixel array.

At first i used the cavemen method of debugging by setting a log message in each function i made lol, but then i used the debugger, it was super slow, but it helped me to get it working (i still wish to never use visual studio again)

I will create a new branch and setup everything from scratch, i wanna do some cleanup and such, then make the pull request from there.

But before that, i need to at least make a proof of concept, a plugin that actually benefits from this so they actually consider accepting this feature

And lastly if you have sometime next week i want you to check what i did (after i commit the changes), because i'm sure i messed up something somewhere

**Comment by d-uzlov** [comment]
Sounds very cool!
I'm happy it works already ^_^

I'll add my thoughts on code and design solutions later.

I have few quick suggestions:
1. Go to project settings in visual studio and check if debug c++ libraries are used. If they are used, disable them. VS has extremely slow debug libraries, and most of the time you don't even need them.
That's why in my plugins' VS projects I have `debug` and `test` build modes: second one disables useless part of the debugging, and is like 10 times faster because of it, while still allowing comfortable debugging.
2. Reset the master branch to official rainmeter state, push everything into a new branch and create a pull request inside your repository. This way it will be easier to track changes, and it will be easier to discuss changes.

**Comment by LI7XI** [comment]
>Reset the master branch to official rainmeter state, push everything into a new branch and create a pull request inside your repository. This way it will be easier to track changes, and it will be easier to discuss changes.

~~ah.. i've been trying for a little while now, how to do this? usually i delete the entire repo and reclone it lol.~~
nvm i got it working

**Comment by LI7XI** [comment]
Btw, it supports transparency as well

![image](https://user-images.githubusercontent.com/55055075/120879290-9f304480-c5c2-11eb-8d69-eeea0a73dc85.png)

**Comment by LI7XI** [comment]
>Now, what do you (as a plugin dev) think of the following?

I noticed that i have a memory leak when calling `UpdatePixels`, even when i delete the array and reallocate it there, idk what causes it.

**Comment by d-uzlov** [comment]
> I noticed that i have a memory leak when calling `UpdatePixels`, even when i delete the array and reallocate it there, idk what causes it.

Is it really in the `UpdatePixels` and not somewhere inside rainmeter? Like, maybe the cache is leaking, or something.

**Comment by LI7XI** [comment]
>Is it really in the UpdatePixels and not somewhere inside rainmeter? Like, maybe the cache is leaking, or something.

I had something in mind but i didn't fix it yet.
[Here](https://github.com/LI7XI/rainmeter/commit/f19516ae2e617fdd2336630f06c261088ad34b86#diff-5aff9a4d04eeada3cb93c708ca18b14f7040c93c481f7063b50f4c57029c11adR312) i instantiate a new struct, but one of its members contains a pointer, and i didn't free the content of that pointer when the function ends, so i get a mem leak because the struct is created every time image meter requests data :P

I will fix it in my next mini refactor.

**Comment by LI7XI** [comment]
@d-uzlov help🥺 that memory leak didn't go away even after the refactoring.

It's not from the plugin, something internally (in `LoadImageFromPluginMeasure` function) is causing this, to make sure i went and commented out `UpdatePixels` function in `Update` function but kept the `timestamp++` part, as expected the image pixels didn't update but internally the image meter reads the same pixel over and over again, and the memory increases rapidly.

How you find/fix memory leaks? also how do you debug plugins? i spent the whole day desperately trying to fix it  without success :(

**Comment by LI7XI** [comment]
>maybe the cache is leaking, or something.

I didn't check this tho, maybe it's related, i'll check it later.

**Reply by LI7XI** [reply]
@d-uzlov i found the cause, it was [this](https://github.com/LI7XI/rainmeter/commit/a94f364b3f18e845b0a0224115acb5dc6f3e89e1#diff-d4d3a5a7c9b8b989f8e4a39ade49f5797250389227d0a52d7d9a5c2177526d54L131), calling `->Release()` fixed the issue :D

**Comment by d-uzlov** [comment]
I don't have access to my PC right now, I'll look into it tomorrow, if you don't find it out before me.

>How you find/fix memory leaks?

Um, I guess I usually just try to disable features one by one to find the culprit.
Also, if you suspect the cache, for example, you can look at its size in debug mode, or just print it.

>also how do you debug plugins

You build them in debug-compatible mode, connect your VS to rainmeter and add a breakpoint where you want.
But in your case you probably want to debug rainmeter itself, so you build the rainmeter in debug mode, and you can actually just launch it via visual studio debug button, because it's just a program.

**Comment by LI7XI** [comment]
>Um, I guess I usually just try to disable features one by one to find the culprit.

Binary search but for debugging XD, got it.

>You build them in debug-compatible mode, connect your VS to rainmeter and add a breakpoint where you want.
>But in your case you probably want to debug rainmeter itself, so you build the rainmeter in debug mode, and you can actually just launch it via visual studio debug button, because it's just a program.

But it says it can't load debug symbols (for my plugin), even if i added the .pdb to plugins folder as well.

**Comment by d-uzlov** [comment]
>But it says it can't load debug symbols (for my plugin), even if i added the .pdb to plugins folder as well.

Maybe VS expects them with a different name. Maybe there are different formats of .pdb, and VS doesn't support yours.
I personally just never had issues with it, so I can't think of any certain solution.

**Comment by LI7XI** [comment]
It's fine, i'm not used to debuggers anyway, `printf` go brrrr.

**Comment by LI7XI** [comment]
If i just found out earlier about turning existing issue into a discussion...

**Comment by d-uzlov** [comment]
Cool. I didn't even knew about discussions :)

**Comment by LI7XI** [comment]
@d-uzlov i spent 4 days just to get these few lines to work, i hate windows api from the bottom of my soul🙂

```cpp
  LPCWSTR imagePath = RmReplaceVariables(rm, RmReadString(rm, L"ImagePath", L""));
  RmLog(rm, LOG_NOTICE, imagePath);

  FILE *file;
  if (_wfopen_s(&file, std::wstring(imagePath).c_str(), L"rb") == 0)
  {
    imgData = stbi_load_from_file(file, &width, &height, &c, 4);
    fclose(file);
  }
```

Aside from that, i have a question, you mentioned that your plugin can process audio on background thread, what is that? and is it also limited by the skin update rate?

I wanna start working on video player plugin, but i got few problems:

UpdateRate of image meter: i can fix this by setting `updateDivider` to -1 and then update it with bangs whenever i need to draw a new frame, my problem with this is that it may use more cpu usage than the normal update (because of bangs), but it's the only way i can think of to get it working "correctly".

UpdateRate of plugin itself: i want to process the video on a new thread so i can control it better with `this_thread::sleep_for()`, the problem is, if any thread that plugins create will be limited by the original skin update rate then i'm screwed, i can't set the update rate to 16 for the entire skin.

Also trying to sync between the frame rate and update rate would be a nightmare, i can somehow bypass this with `defaultUpdateDivider` in rainmeter section but i'm making a plugin, and the users who gonna use it may not get why it's so complicated.

**TL;DR: what is background thread and is it also limited by the skin update rate or it updates itself freely?**

**Reply by d-uzlov** [reply]
Background threads are not limited by rainmeter update rate.
Rainmeter only affects how often your `Update` function of the plugin is called.
